### ISSUE TITLE
Export to speedscope sometimes has asymmetrical open/close frame event sequencing

### ISSUE THREAD

**Issue by jlfwong** [microsoft/perfview#1178]
Hi! I received a bug report to speedscope (https://github.com/jlfwong/speedscope/issues/272) which I'm fairly sure is caused by the event sequencing of the speedscope export of nettrace being incorrect. There's more in-depth discussion of the problem and relevant files in that issue.

If someone does plan on patching this issue, it would also be helpful to update this line:

https://github.com/microsoft/perfview/blob/9ca3b2e803d862935fd29789d550fe97a3bc9b4e/src/TraceEvent/Stacks/SpeedScopeStackSourceWriter.cs#L55

The intent of that field in the file format is to indicate what program exported the file so I can figure out if it's a bug in speedscope's export format or not. In this particular case, it was not, but if it weren't for @swift-kim being very thorough in their bug report, I might've gone down a rabbit hole trying to figure out where the bug was.

I'd prefer if that line was instead:

```
 writer.Write("\"exporter\": \"microsoft/perfview\", "); 
```

possibly with a version number.

**Comment by swift-kim** [comment]
It seems the problem is with sample aggregation in `GetAggregatedOrderedProfileEvents()`. To be more specific, it's kind of precision error in time computation. It is normally expected that:

```
sample.RelativeTime + sample.Metric <= nextSample.RelativeTime
```

but we have cases like (notice the subtle error 0.000000033943):

```
1150.594691 + 1.278308033943 = 1151.872999033943 > 1151.872999
```

As a quick fix, I reduced every Metric's size by 0.0000001 and it worked like a magic. I don't know why RelativeTime and Metric have different precisions. @adamsitnik Do you know any good way to fix?

**Comment by swift-kim** [comment]
@adamsitnik Another related issue with the sample aggregation is broken callstacks observed from Xamarin.Forms' recursive renderer creation. Even with the above quick fix, the resulting speedscope profile looks like:

![image](https://user-images.githubusercontent.com/40190588/82787239-f7594480-9ea0-11ea-953b-2419e7aff099.png)

However if sample aggregation is disabled, the callstacks are ordered correctly.

![image](https://user-images.githubusercontent.com/40190588/82787351-2c659700-9ea1-11ea-8a05-0a742f526442.png)

c.f. Instrumented method tracing (not sampling)

![image](https://user-images.githubusercontent.com/40190588/82788353-44d6b100-9ea3-11ea-878d-0bd37145540c.png)

**Comment by adamsitnik** [comment]
@jlfwong first of all: thank you for the amazing speedscope app!

I've sent a PR that adds the necessary information: https://github.com/microsoft/perfview/pull/1181

I should have included this from the beginning, it would avoid the following question: @swift-kim which version of PerfView|dotnet trace are you using? I've downloaded the trace file that you have provided in https://github.com/jlfwong/speedscope/issues/272 and exported it to speedscope using the latest PerfView version. The order of events is different: https://gist.github.com/adamsitnik/86a982f6b854ceee6ba4481a5e504364

In March I've improved our speedscope exporter (#1102) and I wonder if this PR has already fixed the problem that you are facing.

Thanks,
Adam

**Comment by swift-kim** [comment]
@adamsitnik Thanks for the response and also your great work on the speedscope exporter.

I was using the latest release of dotnet-trace (3.1.120604) so probably I was using TraceEvent 2.0.51 which already includes your patch AFAIK.

I didn't use PerfView to export to the speedscope format but used the `dotnet trace convert --format Speedscope` command to convert it.

I'm not quite sure how you obtained your output but it looks like this:
https://www.speedscope.app/#profileURL=https://gist.githubusercontent.com/adamsitnik/86a982f6b854ceee6ba4481a5e504364/raw/5225c864c97e594d9a7ed1c7d08f27f63e4959a6/includeProductInfo.speedscope.json
in speedscope.app and unfortunately the frames are not aggregated at all.

**Comment by brianrob** [comment]
Looks like this got auto-closed based on #1181.  @adamsitnik can you please take a look at the rest of the issue as well?

**Comment by adamsitnik** [comment]
Now the issue can be closed. Once again thank you @swift-kim and @jlfwong !

**Comment by swift-kim** [comment]
@adamsitnik Ugh. I got another sample trace from one of our developers which unfortunately broke speedscope 1.6.0 again.

In short, it seems the magic number 1.2 is too large for general cases.

In the trace, the following two samples of the same frame number (237) are not continuous, because there are actually two other samples between the two samples. However, they pass the magic number test because of the large metric value (17.8715).
```
Relative Time | Metric
  9825.077032 | 17.8715
  9846.333647 | 22.49046
```
To work around this with minimal effort, can we change the value to something close to 1.0? (I tested 1.1 and it worked.)

Thanks!

[31116.speedscope.zip](https://github.com/microsoft/perfview/files/4728265/31116.speedscope.zip)
